File: src/model/QuizModel.ts
import { Timestamp } from "firebase/firestore";

// ---------------------
// Quiz Question Model
// ---------------------
export interface IQuizQuestionModel {
  questionId: string;
  type: string;
  question: string;
  options?: string[];
  correctAnswer: string;
  explanation?: string;
  points: number;
}

export class QuizQuestionModel implements IQuizQuestionModel {
  questionId: string;
  type: string;
  question: string;
  options?: string[];
  correctAnswer: string;
  explanation?: string;
  points: number;

  constructor(data: IQuizQuestionModel) {
    this.questionId = data.questionId;
    this.type = data.type;
    this.question = data.question;
    this.options = data.options;
    this.correctAnswer = data.correctAnswer;
    this.explanation = data.explanation;
    this.points = data.points;
  }

  static fromFirestore(json: any): QuizQuestionModel {
    return new QuizQuestionModel({
      questionId: json.questionId ?? "",
      type: json.type ?? "multiple_choice",
      question: json.question ?? "",
      options: json.options?.map((o: any) => String(o)),
      correctAnswer: json.correctAnswer ?? "",
      explanation: json.explanation,
      points: json.points ?? 1,
    });
  }

  toFirestore(): any {
    return {
      questionId: this.questionId,
      type: this.type,
      question: this.question,
      options: this.options,
      correctAnswer: this.correctAnswer,
      explanation: this.explanation,
      points: this.points,
    };
  }

  copyWith(params: Partial<IQuizQuestionModel>): QuizQuestionModel {
    return new QuizQuestionModel({
      questionId: params.questionId ?? this.questionId,
      type: params.type ?? this.type,
      question: params.question ?? this.question,
      options: params.options ?? this.options,
      correctAnswer: params.correctAnswer ?? this.correctAnswer,
      explanation: params.explanation ?? this.explanation,
      points: params.points ?? this.points,
    });
  }
}

// ---------------------
// Quiz Model
// ---------------------
export interface IQuizModel {
  id: string;
  title: string;
  description: string;
  levelNumber: number;
  timeLimit?: number;
  passingScore: number;
  allowRetakes: boolean;
  maxAttempts?: number;
  questions: QuizQuestionModel[];
  totalQuestions: number;
  totalPoints: number;
  accessTier: string;
  isActive: boolean;
  sortOrder: number;
  createdAt: Date;
  updatedAt: Date;
  createdBy: string;
  levelId: string;
}

export class QuizModel implements IQuizModel {
  id: string;
  title: string;
  description: string;
  levelNumber: number;
  timeLimit?: number;
  passingScore: number;
  allowRetakes: boolean;
  maxAttempts?: number;
  questions: QuizQuestionModel[];
  totalQuestions: number;
  totalPoints: number;
  accessTier: string;
  isActive: boolean;
  sortOrder: number;
  createdAt: Date;
  updatedAt: Date;
  createdBy: string;
  levelId: string;

  constructor(data: IQuizModel) {
    this.id = data.id;
    this.title = data.title;
    this.description = data.description;
    this.levelNumber = data.levelNumber;
    this.timeLimit = data.timeLimit;
    this.passingScore = data.passingScore;
    this.allowRetakes = data.allowRetakes;
    this.maxAttempts = data.maxAttempts;
    this.questions = data.questions;
    this.totalQuestions = data.totalQuestions;
    this.totalPoints = data.totalPoints;
    this.accessTier = data.accessTier;
    this.isActive = data.isActive;
    this.sortOrder = data.sortOrder;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
    this.createdBy = data.createdBy;
    this.levelId = data.levelId;
  }

  static create(params: {
    id: string;
    title: string;
    description: string;
    levelNumber: number;
    questions: QuizQuestionModel[];
    passingScore: number;
    timeLimit?: number;
    allowRetakes?: boolean;
    maxAttempts?: number;
    accessTier?: string;
    sortOrder?: number;
    createdBy?: string;
    levelId?: string;
  }): QuizModel {
    const now = new Date();
    const totalQ = params.questions.length;
    const totalP = params.questions.reduce((sum, q) => sum + q.points, 0);

    return new QuizModel({
      id: params.id,
      title: params.title,
      description: params.description,
      levelNumber: params.levelNumber,
      timeLimit: params.timeLimit,
      passingScore: params.passingScore,
      allowRetakes: params.allowRetakes ?? true,
      maxAttempts: params.maxAttempts,
      questions: params.questions,
      totalQuestions: totalQ,
      totalPoints: totalP,
      accessTier: params.accessTier ?? "free",
      isActive: true,
      sortOrder: params.sortOrder ?? 0,
      createdAt: now,
      updatedAt: now,
      createdBy: params.createdBy ?? "",
      levelId: params.levelId ?? "",
    });
  }

  static fromFirestore(data: any): QuizModel {
    return new QuizModel({
      id: data.id ?? "",
      title: data.title ?? "",
      description: data.description ?? "",
      levelNumber: data.levelNumber ?? 1,
      timeLimit: data.timeLimit,
      passingScore: (data.passingScore ?? 0) as number,
      allowRetakes: data.allowRetakes ?? true,
      maxAttempts: data.maxAttempts,
      levelId: data.levelId ?? "",
      questions:
        (data.questions as any[])?.map((q) => QuizQuestionModel.fromFirestore(q)) ?? [],
      totalQuestions: data.totalQuestions ?? 0,
      totalPoints: data.totalPoints ?? 0,
      accessTier: data.accessTier ?? "free",
      isActive: data.isActive ?? true,
      sortOrder: data.sortOrder ?? 0,
      createdAt: QuizModel._toDate(data.createdAt),
      updatedAt: QuizModel._toDate(data.updatedAt),
      createdBy: data.createdBy ?? "",
    });
  }

  toFirestore(): any {
    return {
      id: this.id,
      title: this.title,
      description: this.description,
      levelNumber: this.levelNumber,
      timeLimit: this.timeLimit,
      passingScore: this.passingScore,
      allowRetakes: this.allowRetakes,
      maxAttempts: this.maxAttempts,
      questions: this.questions.map((q) => q.toFirestore()),
      totalQuestions: this.totalQuestions,
      totalPoints: this.totalPoints,
      accessTier: this.accessTier,
      isActive: this.isActive,
      sortOrder: this.sortOrder,
      createdAt: Timestamp.fromDate(this.createdAt),
      updatedAt: Timestamp.fromDate(this.updatedAt),
      createdBy: this.createdBy,
      levelId: this.levelId,
    };
  }

  copyWith(params: Partial<IQuizModel>): QuizModel {
    return new QuizModel({
      ...this,
      ...params,
    });
  }

  private static _toDate(value: any): Date {
    if (!value) return new Date();
    if (value instanceof Timestamp) return value.toDate();
    if (value instanceof Date) return value;
    if (typeof value === "string") return new Date(value);
    return new Date();
  }
}

File: Models/quiz model/quiz_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class QuizModel {
  final String id;
  final String title;
  final String description;
  final int levelNumber;
  final int? timeLimit;
  final double passingScore;
  final bool allowRetakes;
  final int? maxAttempts;
  final List<QuizQuestionModel> questions;
  final int totalQuestions;
  final int totalPoints;
  final String accessTier;
  final bool isActive;
  final int sortOrder;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String createdBy;

  const QuizModel({
    required this.id,
    required this.title,
    required this.description,
    required this.levelNumber,
    this.timeLimit,
    required this.passingScore,
    required this.allowRetakes,
    this.maxAttempts,
    required this.questions,
    required this.totalQuestions,
    required this.totalPoints,
    required this.accessTier,
    required this.isActive,
    required this.sortOrder,
    required this.createdAt,
    required this.updatedAt,
    required this.createdBy,
  });

  factory QuizModel.create({
    required String id,
    required String title,
    required String description,
    required int levelNumber,
    required List<QuizQuestionModel> questions,
    required double passingScore,
    int? timeLimit,
    bool allowRetakes = true,
    int? maxAttempts,
    String accessTier = 'free',
    int sortOrder = 0,
    String createdBy = '',
  }) {
    final now = DateTime.now();
    final totalQ = questions.length;
    final totalP = questions.fold<int>(0, (sum, q) => sum + q.points);
    final estimated = (totalQ * 0.75).ceil();
    return QuizModel(
      id: id,
      title: title,
      description: description,
      levelNumber: levelNumber,
      timeLimit: timeLimit,
      passingScore: passingScore,
      allowRetakes: allowRetakes,
      maxAttempts: maxAttempts,
      questions: questions,
      totalQuestions: totalQ,
      totalPoints: totalP,
      accessTier: accessTier,
      isActive: true,
      sortOrder: sortOrder,
      createdAt: now,
      updatedAt: now,
      createdBy: createdBy,
    );
  }

  Map<String, dynamic> toFirestore() => {
    'id': id,
    'title': title,
    'description': description,
    'levelNumber': levelNumber,
    'timeLimit': timeLimit,
    'passingScore': passingScore,
    'allowRetakes': allowRetakes,
    'maxAttempts': maxAttempts,
    'questions': questions.map((q) => q.toFirestore()).toList(),
    'totalQuestions': totalQuestions,
    'totalPoints': totalPoints,
    'accessTier': accessTier,
    'isActive': isActive,
    'sortOrder': sortOrder,
    'createdAt': Timestamp.fromDate(createdAt),
    'updatedAt': Timestamp.fromDate(updatedAt),
    'createdBy': createdBy,
  };

  factory QuizModel.fromFirestore(Map<String, dynamic> json) => QuizModel(
    id: json['id'] ?? '',
    title: json['title'] ?? '',
    description: json['description'] ?? '',
    levelNumber: json['levelNumber'] ?? 1,
    timeLimit: json['timeLimit'],
    passingScore: (json['passingScore'] ?? 0.0).toDouble(),
    allowRetakes: json['allowRetakes'] ?? true,
    maxAttempts: json['maxAttempts'],
    questions:
        (json['questions'] as List<dynamic>?)
            ?.map((e) => QuizQuestionModel.fromFirestore(e))
            .toList() ??
        [],
    totalQuestions: json['totalQuestions'] ?? 0,
    totalPoints: json['totalPoints'] ?? 0,
    accessTier: json['accessTier'] ?? 'free',
    isActive: json['isActive'] ?? true,
    sortOrder: json['sortOrder'] ?? 0,
    createdAt: _toDateTime(json['createdAt']),
    updatedAt: _toDateTime(json['updatedAt']),
    createdBy: json['createdBy'] ?? '',
  );

  QuizModel copyWith({
    String? id,
    String? title,
    String? description,
    int? levelNumber,
    int? timeLimit,
    double? passingScore,
    bool? allowRetakes,
    int? maxAttempts,
    List<QuizQuestionModel>? questions,
    int? totalQuestions,
    int? totalPoints,
    int? estimatedTime,
    String? accessTier,
    bool? isActive,
    int? sortOrder,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? createdBy,
  }) => QuizModel(
    id: id ?? this.id,
    title: title ?? this.title,
    description: description ?? this.description,
    levelNumber: levelNumber ?? this.levelNumber,
    timeLimit: timeLimit ?? this.timeLimit,
    passingScore: passingScore ?? this.passingScore,
    allowRetakes: allowRetakes ?? this.allowRetakes,
    maxAttempts: maxAttempts ?? this.maxAttempts,
    questions: questions ?? this.questions,
    totalQuestions: totalQuestions ?? this.totalQuestions,
    totalPoints: totalPoints ?? this.totalPoints,
    accessTier: accessTier ?? this.accessTier,
    isActive: isActive ?? this.isActive,
    sortOrder: sortOrder ?? this.sortOrder,
    createdAt: createdAt ?? this.createdAt,
    updatedAt: updatedAt ?? this.updatedAt,
    createdBy: createdBy ?? this.createdBy,
  );

  static DateTime _toDateTime(dynamic v) {
    if (v is Timestamp) return v.toDate();
    if (v is String) return DateTime.parse(v);
    return DateTime.now();
  }
}

/* -------------------------------------------------- */
class QuizQuestionModel {
  final String questionId;
  final String type;
  final String question;
  final List<String>? options;
  final String correctAnswer;
  final String? explanation;
  final int points;

  const QuizQuestionModel({
    required this.questionId,
    required this.type,
    required this.question,
    this.options,
    required this.correctAnswer,
    this.explanation,
    required this.points,
  });

  Map<String, dynamic> toFirestore() => {
    'questionId': questionId,
    'type': type,
    'question': question,
    'options': options,
    'correctAnswer': correctAnswer,
    'explanation': explanation,
    'points': points,
  };

  factory QuizQuestionModel.fromFirestore(Map<String, dynamic> json) =>
      QuizQuestionModel(
        questionId: json['questionId'] ?? '',
        type: json['type'] ?? 'multiple_choice',
        question: json['question'] ?? '',
        options: json['options']?.cast<String>(),
        correctAnswer: json['correctAnswer'] ?? '',
        explanation: json['explanation'],
        points: json['points'] ?? 1,
      );

  QuizQuestionModel copyWith({
    String? questionId,
    String? type,
    String? question,
    List<String>? options,
    String? correctAnswer,
    String? explanation,
    int? points,
  }) => QuizQuestionModel(
    questionId: questionId ?? this.questionId,
    type: type ?? this.type,
    question: question ?? this.question,
    options: options ?? this.options,
    correctAnswer: correctAnswer ?? this.correctAnswer,
    explanation: explanation ?? this.explanation,
    points: points ?? this.points,
  );
}
